<fix>[vm]: support memory snapshot

MatheMatrix opened the pull request:
Resolves: ZSV-8948

Change-Id: I6a7a72687971616c6479727061796a7374736a66

sync from gitlab !6113

taogan21 committed:
<fix>[vm]: support memory snapshot

Resolves: ZSV-8948

Change-Id: I6a7a72687971616c6479727061796a7374736a66

taogan21 committed:
<fix>[vm]: define xml after restore from memory snapshot

Resolves: ZSV-9618

Change-Id: I62616a6879727372637476746870797965616b70

taogan21 committed:
<fix>[vm]: open the CD-ROM tray before attaching or detaching iso

Resolves: ZSV-9608

Change-Id: I746f696b626571617071796b64747768786e7469

zstack-robot-2 commented on kvmagent/kvmagent/test/localstorage_testsuite/test_revert_volume_snapshot_group_with_memory_snapshot.py:
```diff
@@ -0,0 +1,299 @@
+                   memorySnapshotUuid: memorySnapshotPath}
+
+        # create snapshot body
+        snapshot_jobs = [
+            snapshot_utils.build_snapshot_job(vm_uuid, virtio1_uuid, virtio1_path, newPath[virtio1_uuid]),
+            snapshot_utils.build_snapshot_job(vm_uuid, virtio2_uuid, virtio2_path, newPath[virtio2_uuid]),
+            snapshot_utils.build_snapshot_job(vm_uuid, virtio_scsi1_uuid, virtio_scsi1_path,
+                                              newPath[virtio_scsi1_uuid]),
+            snapshot_utils.build_snapshot_job(vm_uuid, virtio_scsi2_uuid, virtio_scsi2_path,
+                                              newPath[virtio_scsi2_uuid]),
+            snapshot_utils.build_snapshot_job(vm_uuid, memorySnapshotUuid, env.SNAPSHOT_DIR,
+                                              newPath[memorySnapshotUuid], memory=True)]
+
```
**Comment from tao.gan:**

创建快照不需要指定 `device_id` `use_virtio_scsi`

MatheMatrix commented on kvmagent/kvmagent/test/utils/snapshot_utils.py:
```diff
@@ -59,3 +60,39 @@
     }
 }
+
+take_volumes_snapshots_default_cmd_body = {
+    "snapshotJobs": [],
+    "timeout": 10800,
+    "threadContext": {
+        "task-name": "org.zstack.header.volume.APICreateVolumeSnapshotGroupMsg",
+        "api": "4ac67d731c9f47cf9329d63b4f973e3a"
+    },
+    "threadContextStack": [],
+    "taskContext": {
+        "__messagetimeout__": str(10800 * 1000),
+        "__messagedeadline__": linux.get_current_timestamp() * 1000 + 10800 * 1000,
+    },
+    "kvmHostAddons": {
+        "qcow2Options": " -o cluster_size=2097152 "
+    }
+}
```
**Comment from tao.gan:**

已修改

taogan21 committed:
<fix>[kvmagent]: modify based on AI comment

1 add description for step 6 of reorder_disks
2 case test_revert_volume_snapshot_group_with_memory_snapshot:
Fixed snapshot configuration setup

Resolves: ZSV-8948

Change-Id: I65636d6377736f6b766e7366637167756c627070
